File: sdks/node/test/core/codel.test.ts
// S27 — opt-in CoDel load shedding. Behavioral (timing-based): a slow task at
// concurrency 1 backs the queue up, so later jobs stay above target for a full
// interval and CoDel sheds the stalest ones to the DLQ. The controller algorithm
// itself is unit-tested in Rust (`scheduler::codel`).

import { mkdtempSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { afterEach, expect, it } from "vitest";
import { Queue, type Worker } from "../../src/index";

let worker: Worker | undefined;

afterEach(() => {
  worker?.stop();
  worker = undefined;
});

function newQueue(): Queue {
  const dbPath = join(mkdtempSync(join(tmpdir(), "taskito-codel-")), "queue.db");
  return new Queue({ dbPath });
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

it("rejects non-positive-integer CoDel bounds at configureQueue", () => {
  const queue = newQueue();
  for (const bad of [0, -1, 1.5, Number.NaN, Number.POSITIVE_INFINITY]) {
    expect(() =>
      queue.configureQueue("default", { codel: { targetMs: bad, intervalMs: 30 } }),
    ).toThrow(RangeError);
    expect(() =>
      queue.configureQueue("default", { codel: { targetMs: 30, intervalMs: bad } }),
    ).toThrow(RangeError);
  }
  // A valid pair is accepted.
  queue.configureQueue("default", { codel: { targetMs: 1, intervalMs: 30 } });
});

it("sheds stale jobs to the DLQ under sustained overload", async () => {
  const queue = newQueue();
  queue.configureQueue("default", { codel: { targetMs: 1, intervalMs: 30 } });
  queue.task("slow", async () => {
    await sleep(50);
  });

  const total = 20;
  for (let i = 0; i < total; i++) queue.enqueue("slow");

  worker = queue.runWorker({ concurrency: 1, batchSize: 1 });

  // Poll until every job is accounted for (ran or shed) AND at least one was
  // shed — an aggregate-convergence check, never an instantaneous snapshot.
  const deadline = Date.now() + 25_000;
  let codelDead = 0;
  let stats = await queue.stats();
  while (Date.now() < deadline) {
    const dead = await queue.deadLetters(100);
    codelDead = dead.filter((d) => (d.error ?? "").startsWith("codel:")).length;
    stats = await queue.stats();
    if (stats.completed + stats.dead === total && codelDead >= 1) break;
    await sleep(100);
  }

  expect(codelDead).toBeGreaterThanOrEqual(1);
  // Every dead job is a CoDel drop (the task never throws), and nothing is lost.
  expect(stats.dead).toBe(codelDead);
  expect(stats.completed + stats.dead).toBe(total);
});

File: sdks/java/src/test/java/org/byteveda/taskito/core/CodelTest.java
package org.byteveda.taskito.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.file.Path;
import java.time.Duration;
import java.util.List;
import org.byteveda.taskito.Taskito;
import org.byteveda.taskito.model.DeadJob;
import org.byteveda.taskito.model.QueueStats;
import org.byteveda.taskito.task.Task;
import org.byteveda.taskito.worker.Worker;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;
import org.junit.jupiter.api.io.TempDir;

/**
 * S27 — opt-in CoDel load shedding. The controller algorithm is unit-tested in
 * Rust (`scheduler::codel`); this drives the end-to-end shed path with a slow
 * task at concurrency 1 so the queue backs up and stale jobs are shed to the DLQ.
 */
class CodelTest {

    @Test
    void codelValidatesBounds(@TempDir Path dir) {
        try (Taskito queue =
                Taskito.builder().url(dir.resolve("v.db").toString()).open()) {
            assertThrows(IllegalArgumentException.class, () -> queue.codel("default", 0, 10));
            assertThrows(IllegalArgumentException.class, () -> queue.codel("default", 10, -1));
        }
    }

    @Test
    @Timeout(60)
    void shedsStaleJobsUnderOverload(@TempDir Path dir) throws Exception {
        try (Taskito queue =
                Taskito.builder().url(dir.resolve("codel.db").toString()).open()) {
            queue.codel("default", 1, 30);
            Task<String> slow = Task.of("slow", String.class);

            int total = 20;
            for (int i = 0; i < total; i++) {
                queue.enqueue(slow, "x");
            }

            try (Worker worker = queue.worker()
                    .concurrency(1)
                    .batchSize(1)
                    .handle(slow, payload -> {
                        try {
                            Thread.sleep(50);
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                        }
                        return null;
                    })
                    .start()) {
                // Poll until every job is accounted for and at least one was shed.
                long deadline = System.nanoTime() + Duration.ofSeconds(45).toNanos();
                long codelDead = 0;
                QueueStats stats = queue.stats();
                while (System.nanoTime() < deadline) {
                    List<DeadJob> dead = queue.listDead(100, 0);
                    codelDead = dead.stream()
                            .filter(d -> d.error != null && d.error.startsWith("codel:"))
                            .count();
                    stats = queue.stats();
                    if (stats.completed + stats.dead == total && codelDead >= 1) {
                        break;
                    }
                    Thread.sleep(100);
                }

                // A job can be shed between the dead-letter read and the stats
                // read that satisfied the exit condition; once converged nothing
                // moves, so a fresh dead-letter read gives the settled count.
                codelDead = queue.listDead(100, 0).stream()
                        .filter(d -> d.error != null && d.error.startsWith("codel:"))
                        .count();

                assertTrue(codelDead >= 1, "sustained overload should shed at least one stale job");
                assertEquals(codelDead, stats.dead, "every dead job is a CoDel drop");
                assertEquals(total, stats.completed + stats.dead, "no job is lost");
            }
        }
    }

    @Test
    @Timeout(60)
    void codelConfiguredAfterBuilderIsStillApplied(@TempDir Path dir) throws Exception {
        try (Taskito queue =
                Taskito.builder().url(dir.resolve("late.db").toString()).open()) {
            Task<String> slow = Task.of("slow", String.class);
            int total = 20;
            for (int i = 0; i < total; i++) {
                queue.enqueue(slow, "x");
            }

            // Obtain the builder BEFORE configuring CoDel, then configure it — the
            // late-bound queue-config source must pick this up at start().
            Worker.Builder builder = queue.worker().concurrency(1).batchSize(1).handle(slow, payload -> {
                try {
                    Thread.sleep(50);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
                return null;
            });
            queue.codel("default", 1, 30);

            try (Worker worker = builder.start()) {
                long deadline = System.nanoTime() + Duration.ofSeconds(45).toNanos();
                long codelDead = 0;
                QueueStats stats = queue.stats();
                while (System.nanoTime() < deadline) {
                    codelDead = queue.listDead(100, 0).stream()
                            .filter(d -> d.error != null && d.error.startsWith("codel:"))
                            .count();
                    stats = queue.stats();
                    if (stats.completed + stats.dead == total && codelDead >= 1) {
                        break;
                    }
                    Thread.sleep(100);
                }
                assertTrue(codelDead >= 1, "CoDel set after the builder was obtained must still apply");
            }
        }
    }
}
